src/psyclone/domain/lfric/lfric_stencils.py

```diff
@@ -0,0 +1,649 @@
+        '''
+        root_name = arg.name + "_max_branch_length"
+        unique = LFRicStencils.stencil_unique_str(arg, "length")
+        return symtab.find_or_create_integer_symbol(root_name, tag=unique).name
+
+    def _unique_max_branch_length_vars(self):
+        '''
+        :returns: list of all the unique max stencil extent argument names in
+                  this kernel call for cross2d stencils.
+        :rtype: list of str
+        '''
+        names = []
+        for arg in self._kern_args:
+            if arg.descriptor.stencil['type'] == "cross2d":
+                names.append(arg.name + "_max_branch_length")
+
+        return names
+
+    def _declare_unique_max_branch_length_vars(self, parent):
+        '''
+        Declare all unique max branch length arguments as integers with intent
+        in and add the declaration as a child of the parent argument passed
+        in.
+
+        :param parent: the node in the f2pygen AST to which to add the \
+                       declarations.
+        :type parent: :py:class:`psyclone.f2pygen.SubroutineGen`
+
+        '''
+        api_config = Config.get().api_conf("dynamo0.3")
+
+        if self._unique_max_branch_length_vars():
+            parent.add(DeclGen(
+                parent, datatype="integer",
+                kind=api_config.default_kind["integer"],
+                entity_decls=self._unique_max_branch_length_vars(), intent="in"
+            ))
+
+    @staticmethod
+    def direction_name(symtab, arg):
+        '''
+        Creates a Fortran variable name to hold the direction of the stencil
+        associated with the supplied kernel argument.
+
+        :param symtab: symbol table that will contain (or already contains) \
+            the symbol with this name.
+        :type symtab: :py:class:`psyclone.psyir.symbols.SymbolTable`
+        :param arg: the kernel argument with which the stencil is associated.
+        :type arg: :py:class:`psyclone.dynamo0p3.DynKernelArgument`
+
+        :returns: a Fortran variable name for the stencil direction.
+        :rtype: str
+        '''
+        root_name = arg.name+"_direction"
+        unique = LFRicStencils.stencil_unique_str(arg, "direction")
+        return symtab.find_or_create_integer_symbol(root_name, tag=unique).name
+
+    @property
+    def _unique_extent_vars(self):
+        '''
+        :returns: list of all the unique extent argument names in this \
+                  invoke or kernel call.
+        :rtype: list of str
+
+        :raises InternalError: if neither self._kernel or self._invoke are set.
+
+        '''
+        if self._invoke:
+            names = [arg.stencil.extent_arg.varname for arg in
+                     self._unique_extent_args]
+        elif self._kernel:
+            names = [self.dofmap_size_symbol(self._symbol_table, arg).name
+                     for arg in self._unique_extent_args]
+        else:
+            raise InternalError("_unique_extent_vars: have neither Invoke "
```

Comment:
I would suggest specifying the name of the parent class in this error message to make it easier. That is "LFRicStencils._unique_extent_vars: have ...". The relevant test will need to be adjusted.